src/sdk/glacier_tapecloud.js

```diff
@@ -221,40 +236,116 @@ class TapeCloudGlacier extends Glacier {
+
+                    // Add the enty to the failure log - This could be wasteful as it might
+                    // add entries which have already been migrated but this is a better
+                    // retry.
+                    await failure_recorder(entry);
+                } finally {
+                    await fh?.close(fs_context);
+                }
+            });
+
+            return true;
+        } catch (error) {
+            dbg.error('unexpected error in processing migrate:', error, 'for:', log_file.log_path);
+            return false;
+        }
+    }
+
+    /**
+     * @param {nb.NativeFSContext} fs_context 
+     * @param {LogFile} log_file 
+     * @param {(entry: string) => Promise<void>} failure_recorder 
+     * @returns {Promise<boolean>}
+     */
+    async stage_restore(fs_context, log_file, failure_recorder) {
+        dbg.log2('TapeCloudGlacier.stage_restore starting for', log_file.log_path);
 
-        const file = new LogFile(fs_context, log_file);
         try {
-            await file.collect_and_process(async (entry, batch_recorder) => {
+            await log_file.collect(Glacier.RESTORE_STAGE_WAL_NAME, async (entry, batch_recorder) => {
+                let fh;
                 try {
-                    const should_restore = await Glacier.should_restore(fs_context, entry);
+                    fh = await nb_native().fs.open(fs_context, entry);
+                    const stat = await fh.stat(fs_context);
+
+                    const should_restore = await Glacier.should_restore(fs_context, entry, stat);
                     if (!should_restore) {
                         // Skip this file
                         return;
                     }
 
-                    // Add entry to the tempwal
-                    await batch_recorder(entry);
+                    // If the file staged for migrate then add it to failure log for retry
+                    //
+                    // It doesn't matter if we read the appropriate value for this xattr or not
+                    // 1. If we read nothing we are sure that this file is not staged
+                    // 2. If we read the appropriate value then the file is either being migrated
+                    // or has recently completed migration but hasn't been unmarked.
+                    // 3. If we read corrupt value then either the file is getting staged or is
+                    // getting un-staged - In either case we must requeue.
+                    //
+                    // If we don't find the staging xattr then we must ensure that mtime is
+                    // less than ctime. This is to avoid the case where the file gets
+                    // overwritten hence loses its xattr however migrations are path based
+                    // hence the new file might actually get migrated POST a restore is
+                    // performed.
+                    if (stat.xattr[Glacier.XATTR_STAGE_MIGRATE] || stat.ctime <= stat.mtime) {
```

Comment:
This doesn't work. `stat.ctime <= stat.mtime` will never be true. Remove this. We have a fix for this in `_finalize_restore`.